app/oauth_test.go

```diff
@@ -143,3 +147,355 @@ func TestOAuthDeleteApp(t *testing.T) {
 		t.Fatal("should not get session from cache or db")
 	}
 }
+
+func TestAuthorizeOAuthUser(t *testing.T) {
+	setup := func(enable, tokenEndpoint, userEndpoint bool, serverURL string) *TestHelper {
+		th := Setup()
+
+		th.App.UpdateConfig(func(cfg *model.Config) {
+			cfg.GitLabSettings.Enable = enable
+
+			if tokenEndpoint {
+				cfg.GitLabSettings.TokenEndpoint = serverURL + "/token"
+			} else {
+				cfg.GitLabSettings.TokenEndpoint = ""
+			}
+
+			if userEndpoint {
+				cfg.GitLabSettings.UserApiEndpoint = serverURL + "/user"
+			} else {
+				cfg.GitLabSettings.UserApiEndpoint = ""
+			}
+		})
+
+		return th
+	}
+
+	makeState := func(token *model.Token) string {
+		return base64.StdEncoding.EncodeToString([]byte(model.MapToJson(map[string]string{
+			"token": token.Token,
+		})))
+	}
+
+	makeToken := func(th *TestHelper, cookie string) *model.Token {
+		token, _ := th.App.CreateOAuthStateToken(generateOAuthStateTokenExtra("", "", cookie))
+		return token
+	}
+
+	makeRequest := func(cookie string) *http.Request {
+		request, _ := http.NewRequest(http.MethodGet, "https://mattermost.example.com", nil)
+
+		if cookie != "" {
+			request.AddCookie(&http.Cookie{
+				Name:  COOKIE_OAUTH,
+				Value: cookie,
+			})
+		}
+
+		return request
+	}
+
+	t.Run("not enabled", func(t *testing.T) {
+		th := setup(false, true, true, "")
+		defer th.TearDown()
+
+		_, _, _, err := th.App.AuthorizeOAuthUser(nil, nil, model.SERVICE_GITLAB, "", "", "")
+		require.NotNil(t, err)
+		assert.Equal(t, "api.user.authorize_oauth_user.unsupported.app_error", err.Id)
+	})
+
+	t.Run("with an improperly encoded state", func(t *testing.T) {
+		th := setup(true, true, true, "")
+		defer th.TearDown()
+
+		state := "!"
+
+		_, _, _, err := th.App.AuthorizeOAuthUser(nil, nil, model.SERVICE_GITLAB, "", state, "")
+		require.NotNil(t, err)
+		assert.Equal(t, "api.user.authorize_oauth_user.invalid_state.app_error", err.Id)
+	})
+
+	t.Run("without a stored token", func(t *testing.T) {
+		th := setup(true, true, true, "")
+		defer th.TearDown()
+
+		state := base64.StdEncoding.EncodeToString([]byte(model.MapToJson(map[string]string{
```

Comment:
Can we use `makeState()` here?

Reply:
We'd need to fake a `model.Token` for that, so I don't think we need to do that here.